backend/api/events.py

```diff
@@ -0,0 +1,145 @@
+"""Event API
+
+Event routes are used to create, retrieve, and update Events."""
+
+from fastapi import APIRouter, Depends, HTTPException
+
+from ..services.event import EventNotFoundException, EventService
+from ..services.permission import UserPermissionException
+from ..models.event import Event
+from ..models.event_details import EventDetails
+from ..api.authentication import registered_user
+from ..models.user import User
+
+__authors__ = ['Ajay Gandecha', 'Jade Keegan', 'Brianna Ta', 'Audrey Toney']
+__copyright__ = 'Copyright 2023'
+__license__ = 'MIT'
+
+api = APIRouter(prefix="/api/events")
+openapi_tags = { "name": "Events", "description": "Create, update, delete, and retrieve CS Events."}
+
+@api.get("", response_model=list[EventDetails], tags=['Events'])
+def get_events(event_service: EventService = Depends()) -> list[EventDetails]:
+    """
+    Get all events
+
+    Returns:
+        list[Event]: All `Event`s in the `Event` database table
+    """
+
+    # Return all events
+    return event_service.all()
+
+@api.get("/organization/{slug}", response_model=list[EventDetails], tags=['Events'])
+def get_events_from_organization(slug: str, event_service: EventService = Depends()) -> list[EventDetails]:
+    """
+    Get all events from an organization
+
+    Parameters:
+        slug: a valid str representing a unique Organization
+        event_service: a valid EventService
+
+    Returns:
+        list[EventDetails]: All `EventDetails`s in the `Event` database table from a specific organization
+    """
+
+    # Return all events
+    return event_service.get_events_from_organization(slug)
+
+@api.post("", response_model=EventDetails, tags=['Events'])
+def new_event(event: Event, subject: User = Depends(registered_user), event_service: EventService = Depends()) -> EventDetails:
+    """
+    Create event
+
+    Parameters:
+        event: a valid Event model
+        subject: a valid User model representing the currently logged in User
+        event_service: a valid EventService
+
+    Returns:
+        EventDetails: latest iteration of the created or updated event after changes made
+
+    Raises:
+        HTTPException 404 if create() raises an Exception
+    """
+
+    # Try to create event
+    try:
+        # Return created event
+        return event_service.create(subject, event)
+    except Exception as e:
+        # Raise 422 exception if creation fails
+        # - This would occur if the request body is shaped incorrectly
+        raise HTTPException(status_code=422, detail=str(e))

```

Comment:
@KrisJordan I am curious if this should also use a different Exception, or if we should create one thrown in `event_service` that is handled by the middleware.

Reply:
Yes, ideally it should. The docs also do not reflect the Raises section of the docstring.

Reply:
@KrisJordan would something like InvalidEventException, or potentially more general like InvalidRequestException be okay?